src/main/java/de/tum/in/www1/artemis/repository/metis/ConversationRepository.java

```diff
@@ -45,4 +45,17 @@
              ORDER BY conversation.lastMessageDate DESC
             """)
     List<Conversation> findActiveConversationsOfUserWithConversationParticipants(@Param("courseId") Long courseId, @Param("userId") Long userId);
+
+    @Query("""
+             SELECT COUNT(post.id)
+             FROM ConversationParticipant conversationParticipant
+             JOIN Conversation conversation ON conversation.id = conversationParticipant.conversation.id
+             JOIN Post post ON post.conversation.id = conversation.id
+             WHERE conversation.course.id = :#{#courseId}
+             AND conversation.id = :#{#conversationId}
+             AND conversationParticipant.user.id = :#{#userId}
+             AND post.creationDate > conversationParticipant.lastRead
+            """)
+    long findUnreadMessages(@Param("courseId") Long courseId, @Param("userId") Long userId, @Param("conversationId") Long conversationId);
```

Comment:
If I understand the current logic correctly. Each time any user posts a new message. The `unreadMessages` counter will be always completely new calculated again.  Wouldnt it not make more sense to increase the counter when a new message comes in instead of doing this date range query every single time?

Reply:
Yes, currently new calculation is happening each time. I thought it would be more accurate to calculate and since the query is not costly, it wouldn't cause any performance issues but yes increasing the counter by 1 makes also sense. 

Do you think it is better to increase the counter by 1 instead of query?

Reply:
I am not a database expert but a range query on dates seems far more expensive than simple incrementing a field. Does anybody else have an opinion here?

Reply:
@pellumbbaboci Any idea if you want to change it? is this range query really necessary compared to simply incrementing a counter?

Reply:
@stefanwaldhauser  I will change this to simply increment when sending a new message and decrement when deleting a message. I was wondering if we have to think of many edge cases since we will not calculate each time, but only this two cases come to my mind.

Reply:
changed to simply increment or decrement by 1 :heavy_check_mark: